pbm/backup/physical.go

```diff
@@ -0,0 +1,320 @@
+package backup
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/bsontype"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/x/bsonx/bsoncore"
+
+	"github.com/percona/percona-backup-mongodb/pbm"
+	plog "github.com/percona/percona-backup-mongodb/pbm/log"
+	"github.com/percona/percona-backup-mongodb/pbm/storage"
+)
+
+type Meta struct {
+	ID           UUID                `bson:"backupId"`
+	DBpath       string              `bson:"dbpath"`
+	OplogStart   BCoplogTS           `bson:"oplogStart"`
+	OplogEnd     BCoplogTS           `bson:"oplogEnd"`
+	CheckpointTS primitive.Timestamp `bson:"checkpointTimestamp"`
+}
+
+type BCoplogTS struct {
+	TS primitive.Timestamp `bson:"ts"`
+	T  int64               `bson:"t"`
+}
+
+// see https://www.percona.com/blog/2021/06/07/experimental-feature-backupcursorextend-in-percona-server-for-mongodb/
+type BackupCursorData struct {
+	Meta *Meta
+	Data []pbm.File
+}
+type BackupCursor struct {
+	id    UUID
+	n     *pbm.Node
+	l     *plog.Event
+	close chan struct{}
+}
+
+func NewBackupCursor(n *pbm.Node, l *plog.Event) *BackupCursor {
+	return &BackupCursor{
+		n: n,
+		l: l,
+	}
+}
+
+func (bc *BackupCursor) Data(ctx context.Context) (bcp *BackupCursorData, err error) {
+	cur, err := bc.n.Session().Database("admin").Aggregate(ctx, mongo.Pipeline{
+		{{"$backupCursor", bson.D{}}},
+	})
+	if err != nil {
+		return nil, errors.Wrap(err, "create backupCursor")
+	}
+	defer func() {
+		if err != nil {
+			cur.Close(ctx)
+		}
```

Comment:
Are you sure the cursor should never be closed if any error happens?

Reply:
Here it's gonna be closed if an error happened.
In case of no error, it will be closed [later on](https://github.com/percona/percona-backup-mongodb/pull/638/files/5cb0f474d277c4b2b32d99064425c84021400014#diff-b063b16c36081798de50ca62df2379d8ed16720923203fbb2dc9bcee8b87a3e4R105)